driver/normalizer/normalizer.go

```diff
@@ -772,3 +770,180 @@ func (op dropNils) Check(st *State, n nodes.Node) (bool, error) {
+type opMoveTrivias struct {
+	sub Op
+}
+
+func (op opMoveTrivias) Kinds() nodes.Kind {
+	return nodes.KindObject
+}
+
+func (op opMoveTrivias) Check(st *State, n nodes.Node) (bool, error) {
+	obj, ok := n.(nodes.Object)
+	if !ok {
+		return false, nil
+	}
+	cloned := false
+	leading, ok1 := obj["LeadingTrivia"].(nodes.Array)
+	trailing, ok2 := obj["TrailingTrivia"].(nodes.Array)
+	if ok1 || ok2 {
+		// we saved the leading and trailing trivias, now wipe them from the node
+		obj = obj.CloneObject()
+		cloned = true
+		delete(obj, "LeadingTrivia")
+		delete(obj, "TrailingTrivia")
+	}
+
+	// move comments out of keywords and tokens into actual AST nodes
+	// TODO(dennwc): handle more cases:
+	//				 - Modifiers?
+	for key, sub := range obj {
+		if key != "ReturnType" &&
+			!strings.HasSuffix(key, "Token") &&
+			!strings.HasSuffix(key, "Keyword") {
+			continue
+		}
+		if uast.TypeOf(sub) != typeGroup {
+			continue
+		}
+		group := sub.(nodes.Object)
+		arr, ok := group["Nodes"].(nodes.Array)
+		if !ok {
+			continue
+		}
+		// we already joined trivias and the node into a single array,
+		// so we will have to find its index now
+		ind := -1
+		for i, sub := range arr {
+			if !strings.HasSuffix(uast.TypeOf(sub), "Trivia") {
+				ind = i
+				break
+			}
+		}
+		var node nodes.Node
+		if ind < 0 {
+			// cannot determine an index - pretend everything is a leading trivia
+			// TODO(dennwc): find another way if it happens in practice
+			leading = append(leading.CloneList(), arr...)
+		} else {
+			node = arr[ind]
+			leading = append(leading.CloneList(), arr[:ind]...)
+			trailing = append(arr[ind+1:len(arr):len(arr)], trailing...)
+		}
+		if !cloned {
+			obj = obj.CloneObject()
+			cloned = true
+		}
+		obj[key] = node
+	}
+
+	if len(leading) == 0 && len(trailing) == 0 {
+		if !cloned {
+			return false, nil // unmodified
+		}
+		// removed the trivia fields
+		return op.sub.Check(st, obj)
+	}
+
+	if field, ok := triviaField[uast.TypeOf(obj)]; ok {
+		// instead of wrapping the node, join trivias to a specified field
+		old, ok := obj[field].(nodes.Array)
+		if !ok {
+			return false, fmt.Errorf("expected %q field to be an array", field)
+		}
+		arr := make(nodes.Array, 0, len(leading)+len(old)+len(trailing))
+		arr = append(arr, leading...)
+		arr = append(arr, old...)
+		arr = append(arr, trailing...)
+
+		obj[field] = arr
+	} else {
+		// wrap the node into a uast:Group
+		arr := make(nodes.Array, 0, len(leading)+1+len(trailing))
+		arr = append(arr, leading...)
+		arr = append(arr, obj)
+		arr = append(arr, trailing...)
+
+		// TODO(dennwc): it will be nice if we could extract FullSpan position into the Group
+		group, err := uast.ToNode(uast.Group{})
+		if err != nil {
+			return false, err
+		}
+		obj = group.(nodes.Object)
+		obj["Nodes"] = arr
+	}
+
+	return op.sub.Check(st, obj)
+}
+
+func (op opMoveTrivias) Construct(st *State, n nodes.Node) (nodes.Node, error) {
+	// TODO(dennwc): implement when we will need a reversal
```

Comment:
Can you say a bit more in this comment about what would cause that need?

Reply:
Sorry, I'm too lazy to write down the reasoning behind this each time it happens :)

The original idea behind this pair of functions (`Check` and `Construct`) is that the first checks if the node matches the AST shape that this operation expects, and if so, it populates a "state" with all important pieces of information that this AST node has. This is called a forward pass.

Then, after all the state is collected for this subtree, a reverse pass runs a set of different transforms that use the state to `Construct` the new AST shape. This is the reverse pass mentioned in the comment.

An important note here is that an `Op` should really be used only during a single pass, either forward or reverse. The reasoning behind this is that the "state" that is extracted from AST nodes should be in some kind of canonical form. For example, if a string value is escaped in the AST, we will define an `Unquote` operation that will populate the state with a canonical (unescaped) version of the string. The second pass will use this state directly without using `Unquote` again.

Thus, the `Construct` is only necessary if we want to make the whole pipeline reversible. This is how the whole DSL is designed - we try not to drop state so we will be able to convert AST->UAST and then reverse the pipeline and do UAST->AST conversion. At the end it should allow us to render the source.

But since right now we are trying to make C# sooner rather than later I'm cutting some edges hereand there and skipping the reverse part since we don't really run the pipeline in reverse yet.

More information on the design can be found here: https://github.com/bblfsh/documentation/blob/master/writing-a-driver/transform-dsl.md

Reply:
Ah, sorry, I didn't realize this was a recurrence of an existing pattern. I'd also be fine with a link to an issue, e.g., `TODO(#xyz): implement reversal` or similar. That way I can click through to the context and you don't need to repeat it every time (which, I agree, is a pain).